[MIG] stock_picking_invoice_link: Migration to 15.0

JoelZilli opened the pull request.

eLBati committed:
[add] stock_picking_invoice_link

eLBati committed:
[add] stock_picking_invoice_link

Commit:
[FIX] missing comma in __openerp__.py preventing server to start with a SyntaxError

Commit:
[FIX] missing comma in __openerp__.py preventing server to start with a SyntaxError

eLBati committed:
[ADD] handling _invoice_line_hook

eLBati committed:
[ADD] handling _invoice_line_hook

Commit:
Add one2many from invoice to pickings and display it in the invoice view.
Minor coding style improvements.

Commit:
Add one2many from invoice to pickings and display it in the invoice view.
Minor coding style improvements.

lepistone committed:
autopep8 lint

lepistone committed:
autopep8 lint

Commit:
Port module stock_picking_invoice_link to v8

Commit:
Port module stock_picking_invoice_link to v8

rschnapka committed:
create translation files (template and german)
partially by creating i18n folder first

rschnapka committed:
create translation files (template and german)
partially by creating i18n folder first

jesusVMayor committed:
[ADD]stock_picking_invoice_link: Added spanish translation.

jesusVMayor committed:
[ADD]stock_picking_invoice_link: Added spanish translation.

Commit:
[FIX] removing wrong method _create_invoice_from_picking and values
picking.invoice_id for every stock move.

Commit:
[FIX] removing wrong method _create_invoice_from_picking and values
picking.invoice_id for every stock move.

gurneyalex committed:
Add OCA as author of OCA addons

In order to get visibility on https://www.odoo.com/apps the OCA board has
decided to add the OCA as author of all the addons maintained as part of the
association.

gurneyalex committed:
Add OCA as author of OCA addons

In order to get visibility on https://www.odoo.com/apps the OCA board has
decided to add the OCA as author of all the addons maintained as part of the
association.

eLBati committed:
[FIX] When trying to duplicate an invoice with related pickings you are blocked by this constraint

eLBati committed:
[FIX] When trying to duplicate an invoice with related pickings you are blocked by this constraint

tafaRU committed:
Update copyright year

tafaRU committed:
Update copyright year

rschnapka committed:
OCA Transbot updated translations from Transifex

rschnapka committed:
OCA Transbot updated translations from Transifex

sbidoul committed:
[UPD] prefix versions with 8.0

sbidoul committed:
[UPD] prefix versions with 8.0

pedrobaeza committed:
[MIG] Make modules uninstallable

pedrobaeza committed:
[MIG] Make modules uninstallable

oihane committed:
stock_picking_invoice_link: added new functionality * When an invoice is cancelled its pickings go back to "to2invoiced" state * If a cancelled invoice goes back again to draft, its pickings are marked as invoiced to avoid invoicing twice * Not able to erase an invoice with pickings not cancelled * Tests

stock_picking_invoice_link: unlink improved * enable erasing when invoice is in state 'draft' or 'cancel' * if invoice is in state 'draft', not cancelled picking will be invoiceable

oihane committed:
stock_picking_invoice_link: added new functionality * When an invoice is cancelled its pickings go back to "to2invoiced" state * If a cancelled invoice goes back again to draft, its pickings are marked as invoiced to avoid invoicing twice * Not able to erase an invoice with pickings not cancelled * Tests

stock_picking_invoice_link: unlink improved * enable erasing when invoice is in state 'draft' or 'cancel' * if invoice is in state 'draft', not cancelled picking will be invoiceable

pedrobaeza committed:
stock_picking_invoice_link: Many2many instead of many2one for invoice references

For dropshipping pickings with module stock_dropshipping_dual_invoice, you generate
two invoices for the same picking, so the many2one only reflects the last generated
invoice. With this change, now both are linked.

pedrobaeza committed:
stock_picking_invoice_link: Many2many instead of many2one for invoice references

For dropshipping pickings with module stock_dropshipping_dual_invoice, you generate
two invoices for the same picking, so the many2one only reflects the last generated
invoice. With this change, now both are linked.

alexis-via committed:
[IMP] stock_picking_invoice_link: Refactoring based on new 9.0 invoicing flow

alexis-via committed:
[IMP] stock_picking_invoice_link: Refactoring based on new 9.0 invoicing flow

hendrixcosta committed:
OCA Transbot updated translations from Transifex

hendrixcosta committed:
OCA Transbot updated translations from Transifex

jjscarafia committed:
FIX link invoice view on pickings

jjscarafia committed:
FIX link invoice view on pickings

eLBati committed:
[MIG] stock_picking_invoice_link: Migration to 10.0

avoiding 'product.service_delivery' because not useful in tests and correctly handled as service by 10.0 tests

eLBati committed:
[MIG] stock_picking_invoice_link: Migration to 10.0

avoiding 'product.service_delivery' because not useful in tests and correctly handled as service by 10.0 tests

jbaudoux committed:
[9.0] stock_picking_invoice_link: redo migration from 8.0 to 9.0 (#296)

Previous migration did not keep the business functionality of the previous module. Link from invoice to picking must link only related delivery and not all deliveries in case of partial delivery

jbaudoux committed:
[9.0] stock_picking_invoice_link: redo migration from 8.0 to 9.0 (#296)

Previous migration did not keep the business functionality of the previous module. Link from invoice to picking must link only related delivery and not all deliveries in case of partial delivery

cubells committed:
[IMP] stock_picking_invoice_link: Review changes

cubells committed:
[IMP] stock_picking_invoice_link: Review changes

celm1990 committed:
[MIG] stock_picking_invoice_link: Migration to 11.0

celm1990 committed:
[MIG] stock_picking_invoice_link: Migration to 11.0

chienandalu committed:
[FIX] stock_picking_invoice_link: refund moves

chienandalu committed:
[FIX] stock_picking_invoice_link: refund moves

cubells committed:
[FIX] stock_picking_invoice_link: typo in readme

cubells committed:
[FIX] stock_picking_invoice_link: typo in readme

AdriaGForgeFlow committed:
[12.0][MIG] stock_picking_invoice_link module

AdriaGForgeFlow committed:
[12.0][MIG] stock_picking_invoice_link module

Commit:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (12 of 12 strings)

Translation: stock-logistics-workflow-12.0/stock-logistics-workflow-12.0-stock_picking_invoice_link
Translate-URL: https://translation.odoo-community.org/projects/stock-logistics-workflow-12-0/stock-logistics-workflow-12-0-stock_picking_invoice_link/pt_BR/

Commit:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (12 of 12 strings)

Translation: stock-logistics-workflow-12.0/stock-logistics-workflow-12.0-stock_picking_invoice_link
Translate-URL: https://translation.odoo-community.org/projects/stock-logistics-workflow-12-0/stock-logistics-workflow-12-0-stock_picking_invoice_link/pt_BR/

Commit:
[IMP] stock_picking_invoice_link: black, isort

Commit:
[IMP] stock_picking_invoice_link: black, isort

Commit:
[MIG] stock_picking_invoice_link: Migration to 13.0

Commit:
[MIG] stock_picking_invoice_link: Migration to 13.0

sergio-teruel committed:
[IMP] stock_picking_invoice_link: Allow to update stock move depending on context.

sergio-teruel committed:
[IMP] stock_picking_invoice_link: Allow to update stock move depending on context.

sergio-teruel committed:
[IMP] stock_picking_invoice_link: Better link invoices to stock_moves process.
TT27156

joao-p-marques committed:
[MIG] stock_picking_invoice_link: Migration to 14.0

TT26455

joao-p-marques committed:
[MIG] stock_picking_invoice_link: Migration to 14.0

TT26455

oca-travis committed:
[UPD] Update stock_picking_invoice_link.pot

oca-travis committed:
[UPD] Update stock_picking_invoice_link.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
[UPD] README.rst

pedrocs-exo committed:
Translated using Weblate (Portuguese)

Currently translated at 100.0% (14 of 14 strings)

Translation: stock-logistics-workflow-14.0/stock-logistics-workflow-14.0-stock_picking_invoice_link
Translate-URL: https://translation.odoo-community.org/projects/stock-logistics-workflow-14-0/stock-logistics-workflow-14-0-stock_picking_invoice_link/pt/

pedrocs-exo committed:
Translated using Weblate (Portuguese)

Currently translated at 100.0% (14 of 14 strings)

Translation: stock-logistics-workflow-14.0/stock-logistics-workflow-14.0-stock_picking_invoice_link
Translate-URL: https://translation.odoo-community.org/projects/stock-logistics-workflow-14-0/stock-logistics-workflow-14-0-stock_picking_invoice_link/pt/

MatijaKrolo committed:
Translated using Weblate (Croatian)

Currently translated at 78.5% (11 of 14 strings)

Translation: stock-logistics-workflow-14.0/stock-logistics-workflow-14.0-stock_picking_invoice_link
Translate-URL: https://translation.odoo-community.org/projects/stock-logistics-workflow-14-0/stock-logistics-workflow-14-0-stock_picking_invoice_link/hr/

MatijaKrolo committed:
Translated using Weblate (Croatian)

Currently translated at 78.5% (11 of 14 strings)

Translation: stock-logistics-workflow-14.0/stock-logistics-workflow-14.0-stock_picking_invoice_link
Translate-URL: https://translation.odoo-community.org/projects/stock-logistics-workflow-14-0/stock-logistics-workflow-14-0-stock_picking_invoice_link/hr/

pedrobaeza committed:
[FIX] stock_picking_invoice_link: Read action with sudo

Since v14, actions are not readable by everyone, so we need to read it
with sudo.

Fixes #919

pedrobaeza committed:
[FIX] stock_picking_invoice_link: Read action with sudo

Since v14, actions are not readable by everyone, so we need to read it
with sudo.

Fixes #919

florian-dacosta committed:
[FIX] stock_picking_invoice_link: Read invoice action from picking with sudo

florian-dacosta committed:
[FIX] stock_picking_invoice_link: Read invoice action from picking with sudo

OCA-git-bot committed:
stock_picking_invoice_link 14.0.1.0.1

OCA-git-bot committed:
stock_picking_invoice_link 14.0.1.0.1

antonyht27 committed:
[MIG] stock_picking_invoice_link: Migration to 15.0

antonyht27 committed:
[MIG] stock_picking_invoice_link: Migration to 15.0

celm1990 commented:
@JoelZilli working on https://github.com/OCA/stock-logistics-workflow/pull/951 ping @antonyht27

nicomacr commented:
@celm1990 we continue the migration in this PR to fix the refund part as you said in the comments.
I think this supersede the other PR, do you agree?

celm1990 commented:
> @celm1990 we continue the migration in this PR to fix the refund part as you said in the comments. I think this supersede the other PR, do you agree?

Yes I agree, @antonyht27 please keep in mind this PR for avoid reworking please.

I will be testing PR as soon as possible, thanks for your contribution

kanda999 reviewed (approved):
Functional review: Working as expected.

celm1990 commented on stock_picking_invoice_link/models/account_move.py:
```diff
@@ -0,0 +1,80 @@
+
+    delivery_count = fields.Integer(
+        string="Delivery Orders", compute="_compute_picking_ids"
+    )
```
Any idea for this warning
![image](https://user-images.githubusercontent.com/7775116/173081378-da5fdac6-882f-435c-beec-fc0cf1cfae26.png)

celm1990 commented on stock_picking_invoice_link/models/account_move.py:
```diff
@@ -0,0 +1,80 @@
+            self.move_id.state == "draft"
+            or self.company_currency_id.is_zero(self.amount_cost)
+            or not self.move_line_ids
+        )
```
Please remove this function `_can_update_amount_cost`, is not used here. Keep clean code

celm1990 commented on stock_picking_invoice_link/views/account_invoice_view.xml:
```diff
@@ -0,0 +1,24 @@
+                    icon="fa-truck"
+                    attrs="{'invisible': [('delivery_count', '=', 0)]}"
+                >
+                    <field name="delivery_count" widget="statinfo" string="Pickings" />
```
For consistency between Sale order and invoice

```suggestion
                    <field name="delivery_count" widget="statinfo" string="Delivery" />
```

Sale Order
![image](https://user-images.githubusercontent.com/7775116/173082578-424dbd4b-3a4f-494d-b51b-90022da1de52.png)

Invoice
![image](https://user-images.githubusercontent.com/7775116/173082626-f3c50957-3fe1-4e1d-b6da-e073fc95a692.png)

celm1990 reviewed (changes requested):
@JoelZilli please squash commits

celm1990 reviewed (approved):
Please squash commits

![image](https://user-images.githubusercontent.com/7775116/174926025-cb50df39-8257-4c74-b529-8cb8f774ab7e.png)

rousseldenis commented:
/ocabot migration stock_picking_invoice_link

rousseldenis commented:
@JoelZilli Thanks for this.

Next time, try to improve your PR's title with meaningful statements (e.g.: '[15.0][MIG] stock_picking_invoice_link'). Thanks

rousseldenis commented:
@JoelZilli Could you also follow the migration process explain here : https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-15.0

Especially the Full process part ? Thanks

JoelZilli commented:
@rousseldenis 
Al changes done!  Thanks for the recomendations.

nicomacr reviewed (approved):
LGTM

rousseldenis commented on stock_picking_invoice_link/models/account_move.py:
```diff
@@ -35,7 +40,7 @@ def action_show_picking(self):
         self.ensure_one()
         form_view_name = "stock.view_picking_form"
-        action = self.env.ref("stock.action_picking_tree_all").sudo()
+        action = self.env.ref("stock.action_picking_tree_all")
```
Please use _for_xml_id() instead of read.

rousseldenis replied:
```suggestion
        result = self.env["ir.actions.act_window"]._for_xml_id("stock.action_picking_tree_all")
```

rousseldenis commented on stock_picking_invoice_link/models/account_move.py:
```diff
@@ -56,6 +61,13 @@ class AccountMoveLine(models.Model):
+    def _copy_data_extend_business_fields(self, values):
+        # OVERRIDE to copy the 'move_line_ids' field as well.
+        super()._copy_data_extend_business_fields(values)
+        values["move_line_ids"] = [(6, None, self.move_line_ids.ids)]
```
Could you explain why overriding this ?

yostashiro replied:
Credit note lines wouldn't be linked to stock moves, otherwise. IMO, this part itself is valid.

However the the changes proposed in #750 should also be applied or stock moves get mistakenly regarded as consumed and the subsequent invoice won't be linked to the picking when a full refund is done with credit note (i.e. invoiced quantity of the sales order line is deducted). Maybe this part can be done later as #750 gets processed.

pedrobaeza replied:
This is not correct, as duplicating an invoice will also copy the linked moves. This is covered already in previous versions with latest changes like #1074

yostashiro replied:
> as duplicating an invoice will also copy the linked moves

I don't think this is true as it requires 'include_business_fields' context for this part of the values to be copied, which is not the case with simple duplication. If the changes from #1074 takes care of the links nicely then this part can be argued to be removed as necessary in the port process.

rousseldenis commented on stock_picking_invoice_link/models/sale_order.py:
```diff
@@ -1,6 +1,7 @@
 # Copyright 2013-15 Agile Business Group sagl (<http://www.agilebg.com>)
 # Copyright 2017 Jacques-Etienne Baudoux <je@bcim.be>
 # Copyright 2021 Tecnativa - João Marques
+# Copyright 2022 Antony Herrera - LooErp
```
You should only modify copyrights if your contribution is significant (typically not migrations).

Add yourself to CONTRIBUTORS

rousseldenis replied:
Could you resolve this?

rousseldenis reviewed (changes requested):
Code review.

pedrobaeza committed:
[FIX] stock_picking_invoice_link: Consider DS pickings

Steps to reproduce the problem:

- Create a sales order with Dropshipping route.
- Confirm the sale order and validate the DS picking.
- Invoice the purchase order.
- Invoice the sales order.

Current behavior:

The customer invoice is not linked to the DS picking

Expected behavior:

It's linked.

That's because current algorithm looks for existing invoices linked to
the picking for discarding double links, but it's not having into
account that such invoices can be vendor ones.

We simply filter for customer invoices in the algorithm to fix it.

TT37997

desdelinux reviewed (approved):
LGTM

cc @rousseldenis

ph-obs commented:
Hello everyone. I stumbled upon this pull request. Any idea when it can be merged? Thank you

rousseldenis commented:
@JoelZilli Could you remove lines in copyright where you haven't modified files?

CarlosRoca13 committed:
[IMP] stock_picking_invoice_link: Add test for new use case

CarlosRoca13 committed:
[IMP] stock_picking_invoice_link: Add move_line_ids to account.move view

Adding this field at tree levels we are able to write from methods
and thus use the same technique that is used in this module from
other modules.

JoelZilli commented:
> @JoelZilli Could you remove lines in copyright where you haven't modified files?

Sorry for the delay, i added myself to contributors and removed the copyright lines where the files weren`t modified as you requested.

rousseldenis commented:
@JoelZilli Could you fix pre-commit, tests and rename your commit as it seems a little weird ?

nicomacr commented:
> @JoelZilli Could you fix pre-commit, tests and rename your commit as it seems a little weird ?

@rousseldenis I've made some changes to solve errors with the test, now all green.

rousseldenis commented on stock_picking_invoice_link/models/stock_move.py:
```diff
@@ -1,6 +1,7 @@
 # Copyright 2013-15 Agile Business Group sagl (<http://www.agilebg.com>)
 # Copyright 2015-2016 AvanzOSC
 # Copyright 2016 Pedro M. Baeza <pedro.baeza@tecnativa.com>
+# Copyright 2022 Antony Herrera - LooErp
```
@JoelZilli Missing this

nicomacr replied:
> Some comments and it should be good



> Some comments and it should be good

Ready this changes

rousseldenis commented on stock_picking_invoice_link/models/stock_picking.py:
```diff
@@ -13,6 +14,9 @@ class StockPicking(models.Model):
         comodel_name="account.move", copy=False, string="Invoices", readonly=True
     )
+    invoice_count = fields.Integer(
+        string="Invoices Orders", compute="_compute_invoice_ids"
```
Convention is `_compute_invoice_count`

rousseldenis reviewed (changes requested):
Some comments and it should be good

rousseldenis commented on stock_picking_invoice_link/models/stock_picking.py:
```diff
@@ -0,0 +1,42 @@
+        self.ensure_one()
+        form_view_name = "account.view_move_form"
+        action = self.env.ref("account.action_move_out_invoice_type")
+        result = action.read()[0]
```
Didn't see this one

```suggestion
        result = self.env["ir.actions.act_window"]._for_xml_id("account.action_move_out_invoice_type")
```

nicomacr replied:
Now is fixed

desdelinux commented:
This is ready for merged

@rousseldenis @luisg123v @pedrobaeza @moylop260

pedrobaeza commented on stock_picking_invoice_link/models/account_move.py:
```diff
@@ -16,8 +16,11 @@ class AccountMove(models.Model):
+        help="Related pickings (only when the invoice has been generated from a sale order).",
+    )
+
+    delivery_count = fields.Integer(
```
`picking_count`, as this is also reuse for purchase pickings on module `purchase_stock_picking_invoice_link`.

pedrobaeza commented on stock_picking_invoice_link/models/stock_picking.py:
```diff
@@ -22,12 +25,18 @@ def action_view_invoice(self):
+    @api.depends("invoice_ids")
+    def _compute_invoice_count(self):
+        for order in self:
+            order.invoice_count = len(order.invoice_ids)
```
It's better to use `read_group` for performance reasons.

nicomacr replied:
It's a M2M field, how you suggest to do it?

pedrobaeza replied:
Ah, OK, I see...

pedrobaeza commented:
What about the other change? And CI is red.

CarlosRoca13 committed:
[FIX] stock_picking_invoice_link: Allow link more than one invoice

If we make partial invoices, we need that all partial invoices were linked
to the pickings related.

Steps to reproduce the problem:

1. Create a purchase order buying 10 units of some stockable product
2. Receive all
3. Create new invoice, invoicing just 8 units (This invoice will be linked to the picking)
4. Create new invoice, invoicing the remaining quantity
5. The invoice with 2 units won't be linked to the picking

ValentinVinagre commented:
@JoelZilli will you finish the pr?

CarlosRoca13 committed:
[IMP] stock_picking_invoice_link: Link return to credit note

If the moves linked to invoice has returns linked, when we make
the credit note, we need to link this new refund invoice to the return
moves.

TT39664

rousseldenis commented:
@JoelZilli

CarlosRoca13 commented:
Please add the changes related to this commits:

1. https://github.com/OCA/stock-logistics-workflow/commit/8a472a5703c87465a5c5f3a4873f8b9a12e0c9a9
2. https://github.com/OCA/stock-logistics-workflow/commit/29cfb6e3988a7fcf1e0a5a0fb9bb8508ad826d66

jeroenev commented:
Can you check the failing Unit Tests?

CarlosRoca13 committed:
[IMP] stock_picking_invoice_link: Add support to order invoice policy

CarlosRoca13 commented:
Please fix tests and add this new changes  https://github.com/OCA/stock-logistics-workflow/pull/1135

pedrobaeza commented:
CI is red

pedrobaeza commented:
Superseded by #1147
